tests/test_nblist.py

```diff
@@ -376,7 +380,148 @@ def test_nblist_max_interactions(block_size, tiles):
+)
+def hif2a_single_topology_leg(request):
+    return setup_hif2a_initial_state(request.param)
+
+
+def setup_hif2a_initial_state(host_name: str):
+    forcefield = Forcefield.load_from_file("smirnoff_1_1_0_sc.py")
+    host_config: Optional[HostConfig] = None
+
+    mol_a, mol_b, core = get_hif2a_ligand_pair_single_topology()
+    if host_name == "complex":
+        with resources.path("timemachine.testsystems.data", "hif2a_nowater_min.pdb") as protein_path:
+            host_sys, host_conf, box, _, num_water_atoms = build_protein_system(
+                str(protein_path), forcefield.protein_ff, forcefield.water_ff, mols=[mol_a, mol_b]
+            )
+            box += np.diag([0.1, 0.1, 0.1])  # remove any possible clashes
+        host_config = HostConfig(host_sys, host_conf, box, num_water_atoms)
+    elif host_name == "solvent":
+        host_sys, host_conf, box, _ = build_water_system(4.0, forcefield.water_ff, mols=[mol_a, mol_b])
+        box += np.diag([0.1, 0.1, 0.1])  # remove any possible clashes
+        host_config = HostConfig(host_sys, host_conf, box, host_conf.shape[0])
+    else:
+        assert 0, "Invalid host name"
+
+    st = SingleTopology(mol_a, mol_b, core, forcefield)
+    host = setup_optimized_host(st, host_config)
+
+    lambda_grid = np.array([0.0])
+
+    initial_state = setup_initial_states(st, host, DEFAULT_TEMP, lambda_grid, seed=2024, min_cutoff=None)[0]
+
+    return st, host, host_name, initial_state
+
+
+def compute_mean_tile_density(
+    nblist, frame: NDArray, box: NDArray, cutoff: float, column_block: int = 32, row_block: int = 32
+) -> float:
+    assert column_block == 32
+    assert column_block == row_block
+    N = frame.shape[0]
+    box_diag = np.diagonal(box)
+    assert nblist.get_num_row_idxs() == N
+    ixn_list = nblist.get_nblist(frame, box, cutoff)
+    tile_densities = []
+    num_blocks = (N + row_block - 1) // row_block
+    ixns_per_tile = min(N, column_block) * min(N, row_block)
+
+    row_block_offset = 0
+    # Each ixn_block is the column atoms in the tile that interact with at least one
+    # row block atom
+    for i, ixn_block in enumerate(ixn_list):
+        row_coords = frame[row_block_offset : row_block_offset + row_block]
+        row_coords = np.expand_dims(row_coords, axis=0)
+        column_ixn_atoms = np.array(ixn_block)
+        column_block_offset = 0
+        for _ in range(num_blocks - i):
+            # Find the interactions that would make up a tile, if the tile is empty
+            # don't compute density.
+            tile_column_ixns = column_ixn_atoms[
+                (column_ixn_atoms >= column_block_offset) & (column_ixn_atoms < (column_block_offset + column_block))
+            ]
+            column_block_offset += column_block
+            if tile_column_ixns.size == 0:
+                continue
+            col_coords = np.expand_dims(frame[tile_column_ixns], axis=1)
+            deltas = image_coords(row_coords - col_coords, box_diag)
```

Comment:
identical, but may be clearer to use `delta_r` https://github.com/proteneer/timemachine/blob/ea7af897d3656d1a386c0d0d1d73b1adcdfce832/timemachine/potentials/jax_utils.py#L37

Reply:
Removed `image_coords` and used `delta_r` in https://github.com/proteneer/timemachine/pull/1368/commits/e5857d644f1f110dd44bacea33b54fbeecf11f1c